src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs

```diff
@@ -1090,6 +1120,126 @@ public void UpdateBaseline(long timestampDiff)
                 SliceHolder.SetBaselineToKey(BaselineDate);
             }
 
+            public void TrimSegmentToDate(ref TimeSeriesValuesSegment splitSegment,
+                TimeSeriesValuesSegment readOnlySegment,
+                Span<double> localValues,
+                Span<double> newValues,
+                Span<byte> tagSlice,
+                TimeSeriesValuesSegment.TagPointer localTag,
+                Span<TimestampState> localState,
+                ulong status)
+            {
+                // we ran out of space (up to 2KB) and we need to split the segment. 
+                // last timestamp of current segment == baseline of next segment
+
+                // it's a rare edge case so we move over the current segment again and add values
+                // until encountering a value which its timestamp == segment last timestamp
+                // from that point, we flush the segment and add the leftovers to a new segment
+
+                var originalBaseline = BaselineDate;
+                var nextSegmentBaseline = readOnlySegment.GetLastTimestamp(originalBaseline);
+
+                using (var enumerator = readOnlySegment.GetEnumerator(_context.Allocator))
+                {
+                    while (enumerator.MoveNext(out var localTimestamp, localValues, localState, ref localTag, out var localStatus))
+                    {
+                        var currentLocalTime = originalBaseline.AddMilliseconds(localTimestamp);
+
+                        if (currentLocalTime != nextSegmentBaseline)
+                        {
+                                    var timestampDiff = ((currentLocalTime - originalBaseline).Ticks / 10_000);
+                                    splitSegment.Append(_context.Allocator, (int)timestampDiff, localValues, localTag.AsSpan(), localStatus);
+                                    continue;
+                        }
+
+                        AppendExistingSegment(splitSegment);
+                        splitSegment.Initialize(localValues.Length);
```

Comment:
from this point you need to generate a new segment, right?
Don't you need to update the baseline of the new segment?

Reply:
Yes, we are calling `UpdateBaseline` when we back from `SplitSegment` (line 1074)